week04/code.js

```diff
@@ -0,0 +1,83 @@
+const btns = document.getElementById("btns");
+const start = document.getElementById("start");
+const reset = document.getElementById("reset");
+
+const human = document.getElementById("Hchoice");
+const com = document.getElementById("Cchoice");
+
+const score = document.getElementById("score");
+const result = document.getElementById("result");
+
+const elements = ["✌", "✊", "✋"];
+
+let myChoice = null;
+let Cchoice = null;
+
+let myScore = 0;
+let comScore = 0;
+
+const clickStart = () => {
+  document.querySelectorAll("#btns button").forEach((btn) => {
+    btn.removeAttribute("disabled");
+  });
+  myChoice = null;
+  Cchoice = Math.floor(Math.random() * 3);
+
+  let std = 0;
+  const timer = setInterval(() => {
+    com.innerText = elements[std % 3];
+    std++;
+
+    if (std > 10) {
+      clearInterval(timer);
+      com.innerText = elements[Cchoice];
+      game(myChoice, Cchoice);
+      disableBtn();
+    }
+  }, 300);
+};
+
+const setMyChoice = (choice) => {
+  myChoice = choice;
+  human.innerText = elements[myChoice];
+};
+
+const game = (myChoice, Cchoice) => {
+  console.log("판정 시작!");
+  console.log(myChoice + "/" + Cchoice);
+  if (myChoice === null) {
+    //패
+    human.innerText = "👎";
+    result.innerText = "👎시간 초과로 패배👎";
+    comScore++;
+  } else if (myChoice === Cchoice) {
+    //무
+    result.innerText = "무승부!";
+  } else if ((myChoice + 1) % 3 === Cchoice) {
+    //패
+    result.innerText = "패배😭";
+    comScore++;
+  } else if ((myChoice + 2) % 3 === Cchoice) {
+    //승
+    result.innerText = "🤩승리🤩";
+    myScore++;
+  } else {
+    //오류
+    result.innerText = "오류 발생!!";
+  }
+  score.innerHTML = myScore + " : " + comScore;
+};
+
+const resetAll = () => {
+  console.log("초기화");
+  myScore = 0;
+  comScore = 0;
+  score.innerHTML = myScore + " : " + comScore;
+  result.innerText = "도전을 누르고 3초 안에 선택하자!";
+};
+
+const disableBtn = () => {
+  document.querySelectorAll("#btns button").forEach((btn) => {
+    btn.setAttribute("disabled", "disabled");
+  });
+};
```

Comment:
```suggestion
};

start.addEventListener("click", clickStart);
reset.addEventListener("click", resetAll);
[scissor, rock, paper].map((el, idx) =>
	el.addEventListener("click", () => setMyChoice(idx))
);
```

위와 같이 addEventListener를 사용해서 함수들을 바인딩해줍니다.
이때 반복되는 코드인 가위, 바위, 보 버튼에 대한 코드는 map 함수를 사용해서 반복문으로 바인딩해줬습니다.
map 함수의 첫번째 파라미터인 el은 `[scissor, rock, paper]` 배열 안의 각 요소인 선택한 요소 노드를,
두번째 파라미터인 idx는 배열 요소의 인덱스(0부터 시작)를 반환합니다.

이때 인자(매개변수)를 넘겨받지 않는 clickStart, resetAll 함수와 인자로 0, 1, 2 중 하나의 숫자를 넘겨줘야 하는 setMyChoice 함수를 바인딩하는 방식에 차이가 있는 이유는 다음과 같습니다.
이벤트 핸들러에 함수를 전달할 때, `() => handler(value)`와 같이 화살표 함수를 사용하지 않고 직접 함수를 호출하여 인자를 전달할 수 없습니다. `scissor.addEventListener("click", setMyChoice(idx));`와 같이 작성하면 이벤트가 발생할 때 함수가 즉시 호출되고 그 반환값이 이벤트 핸들러로 전달되기 때문에 예상하는 대로 동작하지 않습니다. 따라서 함수가 바로 호출되지 않게 하려면 반드시 화살표 함수를 이용하여 바인딩해야 합니다.